control-plane/agents/core/src/volume/specs.rs

```diff
@@ -0,0 +1,730 @@
+        specs.nexuses.get(id).cloned()
+    }
+    async fn get_volume(
+        &self,
+        id: &VolumeId,
+    ) -> Option<Arc<Mutex<VolumeSpec>>> {
+        let specs = self.read().await;
+        specs.volumes.get(id).cloned()
+    }
+    // we could also get the replicas from the volume nexuses
+    async fn get_volume_replicas(
+        &self,
+        id: &VolumeId,
+    ) -> Vec<Arc<Mutex<ReplicaSpec>>> {
+        let mut replicas = vec![];
+        let specs = self.read().await;
+        for replica in specs.replicas.values() {
+            let spec = replica.lock().await;
+            if spec.owners.owned_by(id) {
+                replicas.push(replica.clone());
+            }
+        }
+        replicas
+    }
+    async fn get_replica_node(
+        registry: &Registry,
+        replica: &ReplicaSpec,
+    ) -> Option<NodeId> {
+        let pools = registry.get_pools_inner().await.unwrap();
+        pools.iter().find_map(|p| {
+            if p.id == replica.pool {
+                Some(p.node.clone())
+            } else {
+                None
+            }
+        })
+    }
+    // we could also tag the volume with the "latest" nexuses
+    async fn get_volume_nexuses(
+        &self,
+        id: &VolumeId,
+    ) -> Vec<Arc<Mutex<NexusSpec>>> {
+        let mut nexuses = vec![];
+        let specs = self.read().await;
+        for nexus in specs.nexuses.values() {
+            let spec = nexus.lock().await;
+            if spec.owner.as_ref() == Some(id) {
+                nexuses.push(nexus.clone());
+            }
+        }
+        nexuses
+    }
+
+    pub(super) async fn create_nexus(
+        &self,
+        registry: &Registry,
+        request: &CreateNexus,
+    ) -> Result<Nexus, SvcError> {
+        let node = registry.get_node_wrapper(&request.node).await.context(
+            NodeNotFound {
+                node_id: request.node.clone(),
+            },
+        )?;
+
+        let nexus_spec = {
+            let mut specs = self.write().await;
+            if let Some(spec) = specs.get_nexus(&request.uuid) {
+                {
+                    let mut nexus_spec = spec.lock().await;
+                    if nexus_spec.updating {
+                        // already being created
+                        return Err(SvcError::Conflict {});
+                    } else if nexus_spec.state.creating() {
+                        // this might be a retry, check if the params are the
+                        // same and if so, let's retry!
+                        if nexus_spec.ne(request) {
+                            // if not then we can't proceed
+                            return Err(SvcError::Conflict {});
+                        }
+                    } else {
+                        return Err(SvcError::AlreadyExists {
+                            kind: ResourceKind::Nexus,
+                            id: request.uuid.to_string(),
+                        });
+                    }
```

Comment:
This pattern seems to be used quite a lot. I wonder if we can split it out into a function.

Reply:
Yeah I think so